src/main/java/net/minecraftforge/datafix/ForgeSchema.java

```diff
@@ -0,0 +1,340 @@
+        //And all newer versions will have this type.
+        //Due to the way we set up the schemas with events later, that is not possible anymore.
+        //Modders will need to register their types on all versions that they should exist in,
+        //not just in the oldest version.
+
+        if (wrapped != null)
+        {
+            wrapped.registerTypes(schema, entityTypes, blockEntityTypes);
+        }
+        else if (parent != null)
+        {
+            parent.registerTypes(schema, entityTypes, blockEntityTypes);
+        }
+    }
+
+    @Override
+    public Map<String, Supplier<TypeTemplate>> registerEntities(final Schema schema)
+    {
+        //This is a bit wonky, in vanilla it suffices to register the types to one single version
+        //And all newer versions will have this type.
+        //Due to the way we set up the schemas with events later, that is not possible anymore.
+        //Modders will need to register their types on all versions that they should exist in,
+        //not just in the oldest version.
+
+        if (wrapped != null)
+        {
+            return wrapped.registerEntities(schema);
+        }
+        if (parent != null)
+        {
+            return parent.registerEntities(schema);
+        }
+
+        return Maps.newHashMap();
+    }
+
+    @Override
+    public Map<String, Supplier<TypeTemplate>> registerBlockEntities(final Schema schema)
+    {
+        //This is a bit wonky, in vanilla it suffices to register the types to one single version
+        //And all newer versions will have this type.
+        //Due to the way we set up the schemas with events later, that is not possible anymore.
+        //Modders will need to register their types on all versions that they should exist in,
+        //not just in the oldest version.
+
+        if (wrapped != null)
+        {
+            return wrapped.registerBlockEntities(schema);
+        }
+        if (parent != null)
+        {
+            return parent.registerBlockEntities(schema);
+        }
+
+        return Maps.newHashMap();
+    }
+
+    /**
+     * Registers a new type to the Schema with its type template.
+     *
+     * @param recursive Indicates if the type is potentially recursive.
+     * @param type The type in question.
+     * @param template Its template.
+     */
+    @Override
+    public void registerType(final boolean recursive, final DSL.TypeReference type, final Supplier<TypeTemplate> template)
+    {
+        if (TYPE_TEMPLATES == null)
+            TYPE_TEMPLATES = Maps.newHashMap();
+
+        TYPE_TEMPLATES.put(type.typeName(), template);
+        if (recursive && !RECURSIVE_TYPES.containsKey(type.typeName()))
+        {
+            RECURSIVE_TYPES.put(type.typeName(), RECURSIVE_TYPES.size());
+        }
+    }
+
+    /**
+     * Resets the schema so that the type map can be properly rebuild.
+     */
+    public void resetSchema()
```

Comment:
I feel like this potentially should be package private? (Technically it doesn't matter given the class itself is package private but still.)